feat(plugin): add clearURLs

adryd325 opened the pull request:
At the moment this.onSend() doesn't get called for some reason; @Cynosphere and I couldn't figure out why.

I'd like to wait for whatever is causing this to get fixed, whether in my extension or in Vencord

Update: addPreSendListener is just broken, i'll fix that in this PR i guess

adryd325 commented:
Forgot to include dependency

Vendicated commented:
> At the moment this.onSend() doesn't get called for some reason; @ Cynosphere and I couldn't figure out why.

Was the problem just not having the message api enabled? Or is it actually broken

adryd325 commented:
The problem was not having the API enabled
The simplemarkdown patch was making me think unindent was working fine and it had cyn and i confused

Vendicated commented:
Cool! I think that plugin was made before dependencies were a thing and I forgot to update it

If I understood the rules correctly, the ones without @ apply to any URLs and the ones with an @ are paramToRemove@host? It would be helpful to add a little comment explaining how they work for people who want to modify them in the future

It would also be nice to add Spotify to the default rules, the parameters to remove would be si and utm_* (there's a few, I only know utm_source off the top of my head)

Another useful feature would be to have URLs automatically cleaned before opening them

Also, Plugin Settings will be a thing soon™

adryd325 commented:
The rules format was inspired by https://github.com/Smile4ever/Neat-URL

I just added wildcard matching to params and added documentation on how to write rules

botatooo commented:
what about https://docs.clearurls.xyz/1.23.0/specs/rules/

adryd325 commented:
> what about https://docs.clearurls.xyz/1.23.0/specs/rules/

I didn't feel like implementing the entirety of clearurls

adryd325 commented:
this should be ready for merge

Vendicated commented on src/plugins/clearURLs/index.ts:
```diff
@@ -0,0 +1,129 @@
+                hostRule = new RegExp(
+                    "^(www\\.)?" +
+                        splitRule[1]
+                            .replace(/[#-}]/g, '\\$&')
```
did you mean to escape this `-`? otherwise this is the charrange # to } which is a lot
![image](https://user-images.githubusercontent.com/45497981/194075645-54f3c212-38e9-41fc-9e3f-debd96e65f0f.png)

adryd325 replied:
this is what happens when I grab an escape regex from stackoverflow instead of somewhere sensible like lodash lmao

shoud be fixed now

Vendicated commented:
Seems to work fine now! I would push a few changes (mainly formatting & typescript types stuff) if that's okay, then merge

Vendicated commented:
Thank you!
